apps/minifront/src/state/swap.ts

```diff
@@ -32,11 +33,32 @@ import { getAmountFromValue, getAssetIdFromValue } from '@penumbra-zone/getters/
 import { Amount } from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/num/v1/num_pb';
 import { divideAmounts } from '@penumbra-zone/types/amount';
 import { amountMoreThanBalance } from './send';
+import { bech32mAssetId } from '@penumbra-zone/bech32m/passet';
+
+const getMetadataByAssetId = async (
+  traces: SwapExecution_Trace[] = [],
+): Promise<Record<string, Metadata>> => {
+  const map: Record<string, Metadata> = {};
+
+  for (const trace of traces) {
+    for (const value of trace.value) {
```

Comment:
Probably not a huge performance buff, but you could do:

```typescript
 const promises = traces.flatMap(trace =>
    trace.value.map(async value => {
      if (!value.assetId || map[bech32mAssetId(value.assetId)]) return;

      const { denomMetadata } = await viewClient.assetMetadataById({ assetId: value.assetId });

      if (denomMetadata) {
        map[bech32mAssetId(value.assetId)] = denomMetadata;
      }
    }),
  );

  await Promise.all(promises);
```

Reply:
Ah, good call — fixed.